add page for LASTWITHTIME (#246)
pinot-contrib/pinot-docs: configuration-reference/functions/firstwithtime.md

---
description: This section contains reference documentation for the firstwithtime function.
---

# firstwithtime

> Get the first value of `dataColumn` where the `timeColumn` is used to define the time of `dataColumn` and the `dataType` specifies the type of `dataColumn`, which can be `BOOLEAN`, `INT`, `LONG`, `FLOAT`, `DOUBLE`, `STRING`

Reviewer: Looking at the [code](https://github.com/apache/pinot/blob/e08e66c3991b40f0d95c48b9436b97d406fa3311/pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/FirstWithTimeAggregationFunction.java#L37C1-L47C4) first two arguments to the function are expressions and third is a literal specifying the return type of dataColumn. 
Help snippet from code:
`/**
 * This function is used for FirstWithTime calculations.
 * <p>The function can be used as FirstWithTime(dataExpression, timeExpression, 'dataType')
 * <p>Following arguments are supported:
 * <ul>
 *   <li>dataExpression: expression that contains the column to be calculated first on</li>
 *   <li>timeExpression: expression that contains the column to be used to decide which data is first, can be any
 *   Numeric column</li>
 *   <li>dataType: the data type of data column</li>
 * </ul>
 */`
 
 This seems more easier to understand definition vs current one.

Reviewer 2: `timeColumn` can be of type `TIMESTAMP`, `INT`, `LONG`

> Get the first value of `dataColumn` where the `timeColumn` is used to define the time of `dataColumn` and the `dataType` specifies the type of `dataColumn`, which can be `BOOLEAN`, `INT`, `LONG`, `FLOAT`, `DOUBLE`, `STRING`

Reviewer 2: I feel it is a little bit more clear this way: Get the value of `dateColumn` with the smallest `timeColumn` value...

## Signature

> FIRSTWITHTIME(dataColumn, timeColumn, 'dataType')

## Usage Example

This example is based on the [Streaming Quick Start](../../basics/getting-started/quick-start.md#streaming).

```sql

> select FIRSTWITHTIME(group_name, __metadata$recordTimestamp, 'STRING')

Reviewer: Looking at defined schema column used for timeExpression is actually a string 
    {
      "dataType": "STRING",
      "name": "__metadata$recordTimestamp"
    }
and not numeric per code definition, however I am guessing this query runs fine? If yes then perhaps internally we do implicit String to numeric conversion but then more user relatable e.g could be just using time col. for the timeExpression `mTime`. Thoughts?

---
description: This section contains reference documentation for the firstwithtime function.
---

# firstwithtime

> This function is used for FirstWithTime calculations.

Reviewer 3: ```suggestion
```

> The function will return the value of `dataColumn` with the smallest `timeColumn` value where:

Reviewer 3: ```suggestion
Returns the value of `dataColumn` with the smallest `timeColumn` value where:
```

- `timeColumn` is used to define the time of `dataColumn`, which can be of type `TIMESTAMP`, `INT`, `LONG`
- `dataType` specifies the type for `dataColumn`, which can be `BOOLEAN`, `INT`, `LONG`, `FLOAT`, `DOUBLE`, `STRING`

## Signature

> FIRSTWITHTIME(dataColumn, timeColumn, 'dataType')

> ## Usage Example

Reviewer 3: ```suggestion
## Example
```
